chips/nrf52/src/i2c.rs

```diff
@@ -0,0 +1,446 @@
+
+/// An I2C master device.
+///
+/// A `TWIM` instance wraps a `registers::TWIM` together with
+/// addition data necessary to implement an asynchronous interface.
+pub struct TWIM {
+    registers: *const registers::TWIM,
+    nvic_idx: NvicIdx,
+    client: Cell<Option<&'static hil::i2c::I2CHwMasterClient>>,
+    busy: Cell<bool>,
+    buf: TakeCell<'static, [u8]>,
+    enabled: Cell<bool>,
+}
+
+impl TWIM {
+    const fn new(instance: usize) -> TWIM {
+        TWIM {
+            registers: registers::INSTANCES[instance].0,
+            nvic_idx: registers::INSTANCES[instance].1,
+            client: Cell::new(None),
+            busy: Cell::new(false),
+            buf: TakeCell::empty(),
+            enabled: Cell::new(false),
+        }
+    }
+
+    pub fn set_client(&self, client: &'static hil::i2c::I2CHwMasterClient) {
+        debug_assert!(self.client.get().is_none());
+        self.client.set(Some(client));
+    }
+
+    fn regs(&self) -> &registers::TWIM {
+        unsafe { &*self.registers }
+    }
+
+    fn enable_nvic(&self) {
+        nvic::enable(self.nvic_idx);
+    }
+
+    /// Configures an already constructed `TWIM`.
+    pub fn configure(&self, scl: Pinmux, sda: Pinmux) {
+        let regs = self.regs();
+        regs.psel_scl.set(scl);
+        regs.psel_sda.set(sda);
+        self.enable_nvic();
+    }
+
+    /// Enables hardware TWIM peripheral.
+    pub fn enable(&self) {
+        let regs = self.regs();
+        regs.enable.set(registers::Enable::Enabled);
+        self.enabled.set(true);
+    }
+
+    /// Disables hardware TWIM peripheral.
+    pub fn disable(&self) {
+        let regs = self.regs();
+        regs.enable.set(registers::Enable::Disabled);
+        self.enabled.set(false);
+    }
+
+    pub fn handle_interrupt(&self) {
+        if self.regs().events_stopped.get() == 1 {
+            self.regs().events_stopped.set(0);
+            match self.client.get() {
+                None => (),
+                Some(client) => {
+                    match self.buf.take() {
+                        None => (),
+                        Some(buf) => {
+                            client.command_complete(buf, hil::i2c::Error::CommandComplete);
+                        }
+                    }
+                }
+            };
+        }
+
+        if self.regs().events_error.get() == 1 {
+            self.regs().events_error.set(0);
+            let errorsrc = self.regs().errorsrc.get();
+            self.regs().errorsrc.set(registers::ErrorSrc::None);
+            match self.client.get() {
+                None => (),
+                Some(client) => {
+                    match self.buf.take() {
+                        None => (),
+                        Some(buf) => {
+                            client.command_complete(buf, errorsrc.into());
+                        }
+                    }
+                }
+            };
+        }
+
+        if self.regs().events_suspended.get() == 1 {
```

Comment:
Am I reading this correctly that these checks aren't actually doing any work right now, but are here as a template in case we handle those events in the future?

Reply:
@Vagabond can probably answer this better. But I believe the intention is to clear the events event if we aren't monitoring them.

Reply:
Just to clarify, I meant the if statement. We could clear the events without checking if they were active.

Reply:
Yeah, I think we could just blindly clear them here.

Reply:
K, I don't think I particularly care, just wanted to make sure I understood what was going on.